File: packages/sanbox/src/seatbelt.ts
import type { SandboxPolicy } from './types';
import { validatePolicy } from './policy';
import runtime from './runtime-policy.json';

/** SBPL strings require escaping independently of shell/JSON argument transport. */
function quote(path: string): string {
    return `"${path.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`;
}

export function createSeatbeltProfile(policy: SandboxPolicy): string {
    validatePolicy(policy);
    const read = [...new Set([...runtime.seatbelt.runtimePaths, ...policy.readOnlyPaths ?? [], ...policy.writablePaths ?? []])];
    const rules = [...runtime.seatbelt.rules];
    for (const path of read) rules.push(`(allow file-read* file-map-executable (subpath ${quote(path)}))`);
    for (const path of new Set(policy.writablePaths ?? [])) rules.push(`(allow file-write* (subpath ${quote(path)}))`);
    if (policy.network === 'allow') rules.push('(allow network*)');
    return rules.join('\n');
}

File: packages/sanbox/native/src/seatbelt.rs
use std::process::Command;
use crate::{Mount, NetworkAccess, SeatbeltPolicy};

fn quote(path: &str) -> String {
    format!("\"{}\"", path.replace('\\', "\\\\").replace('"', "\\\""))
}

pub(crate) fn command(script: &str, cwd: &str, mounts: &[Mount], network: NetworkAccess, policy: &SeatbeltPolicy) -> Command {
    let mut rules = policy.rules.clone();
    for path in policy.runtime_paths.iter().chain(mounts.iter().map(|mount| &mount.source)) {
        rules.push(format!("(allow file-read* file-map-executable (subpath {}))", quote(path)));
    }
    for mount in mounts.iter().filter(|mount| mount.writable) {
        rules.push(format!("(allow file-write* (subpath {}))", quote(&mount.source)));
    }
    if network == NetworkAccess::Allow { rules.push("(allow network*)".into()); }
    let mut result = Command::new("/usr/bin/sandbox-exec");
    result.args(["-p", &rules.join("\n"), "/bin/bash", "--noprofile", "--norc", "-c", script]).current_dir(cwd);
    result
}
